dace/codegen/targets/framecode.py

```diff
@@ -467,6 +393,317 @@ def dispatch_state(state: SDFGState) -> str:
+                return TOP_SCHEDULE
+            return (sdfg.parent_nsdfg_node.schedule or TOP_SCHEDULE)
+        else:
+            raise TypeError
+
+    def _can_allocate(self, sdfg: SDFG, state: SDFGState, desc: data.Data,
+                      scope: Union[nodes.EntryNode, SDFGState, SDFG]) -> bool:
+        schedule = self._get_schedule(scope)
+        # if not dtypes.can_allocate(desc.storage, schedule):
+        #     return False
+        if dtypes.can_allocate(desc.storage, schedule):
+            return True
+
+        # Check for device-level memory recursively
+        node = scope if isinstance(scope, nodes.EntryNode) else None
+        cstate = scope if isinstance(scope, SDFGState) else state
+        csdfg = scope if isinstance(scope, SDFG) else sdfg
+
+        if desc.storage in dtypes.FPGA_STORAGES:
+            return sdscope.is_devicelevel_fpga(csdfg, cstate, node)
+        elif desc.storage in dtypes.GPU_STORAGES:
+            return sdscope.is_devicelevel_gpu(csdfg, cstate, node)
+
+        # return True
+        return False
+
+    def determine_allocation_lifetime(self, top_sdfg: SDFG):
+        """ 
+        Determines where (at which scope/state/SDFG) each data descriptor
+        will be allocated/deallocated.
+        :param top_sdfg: The top-level SDFG to determine for.
+        """
+        # Gather shared transients
+        shared_transients = {}
+        for sdfg in top_sdfg.all_sdfgs_recursive():
+            shared_transients[sdfg.sdfg_id] = sdfg.shared_transients(
+                check_toplevel=False)
+
+        for sdfg, name, desc in top_sdfg.arrays_recursive():
+            if not desc.transient:
+                continue
+
+            # NOTE: In the code below we infer where a transient should be
+            # declared, allocated, and deallocated. The information is stored
+            # in the `to_allocate` dictionary. The key of each entry is the
+            # scope where one of the above actions must occur, while the value
+            # is a tuple containing the following information:
+            # 1. The SDFG object that containts the transient.
+            # 2. The State id where the action should (approx.) take place.
+            # 3. The Access Node id of the transient in the above State.
+            # 4. True if declaration should take place, otherwise False.
+            # 4. True if allocation should take place, otherwise False.
+            # 4. True if deallocation should take place, otherwise False.
+
+            # Possibly confusing control flow below finds the first state
+            # and node of the data descriptor, or continues the
+            # arrays_recursive() loop
+            first_state_instance: int = None
+            first_node_instance: nodes.AccessNode = None
+            last_state_instance: int = None
+            last_node_instance: nodes.AccessNode = None
+            first = True
+            for id, state in enumerate(sdfg.nodes()):
```

Comment:
this should be a topological sort or DFS of the states. otherwise after transformations state IDs may be reversed or the order does not match the appearance order.

Reply:
I can actually see more potential problems here regarding the topological sort, for example, multiple paths. This would necessitate the use of realloc (with which we cannot use `DACE_ALIGN`?) but there all sorts of other issues if we use it. I will update the codegen to work with SDFGs similar to the example provided in the test but I think we should leave a more elaborate realloc-solution for a different PR. We already support a strict superset of SDFGs compared to master.

Reply:
ok, please leave a comment in the code then, so that we know something better needs to be done. Also, even if you reverse the states (so that they are in the same order as their ids) it seems to fail for an unrelated reason. Does it fail for you as well?

Reply:
This is probably related to https://github.com/spcl/dace/pull/434/commits/360d9a6b08ef40c9c195c38415607bb19f070831